### DISCUSSION TITLE
How to type the `Tab` generic component w/o `as` prop

### DISCUSSION THREAD

**Discussion by jasikpark** [ariakit/ariakit#2813] [Questions] [answered]
I'm trying to pass a ref to the `Tab` component in my `TabLink` implementation (https://ariakit.org/examples/tab-react-router)

But without `as="a"`, `Tab` will only accept a ref for a button element, rather than an anchor element.

I'm trying to upgrade to `render={}` to avoid the deprecation of `as`, but this is an issue for me.

I *could* just pass the ref directly to the `Link`, but then presumably I wouldn't get the auto ref merging.

```tsx
import { Tab } from '@ariakit/react/tab';
import cn from 'classnames';
import { useLayoutEffect, useRef } from 'react';
import { Link } from 'react-router-dom';
import { useHref } from '@hooks/useHref';
import { useRect } from '@hooks/useRect';
import { useAnimatedTabs } from './context/useAnimatedTabs';
import TabStyles from './TabLink.module.css';

type Props = React.ComponentPropsWithRef<typeof Link>;

export function TabLink({ className, to, ...props }: Props) {
  const href = useHref(to);

  // get the store and style changing function from context
  const { store, setActiveRect } = useAnimatedTabs();
  const selectedId = store.useState('selectedId');

  // measure the size of our element, only listen to rect if active
  const ref = useRef<HTMLAnchorElement>(null);

  const isSelected = !!selectedId && ref.current?.href.includes(selectedId);

  const rect = useRect(ref, { observe: isSelected });

  // callback to set styles whenever we're active
  useLayoutEffect(() => {
    if (isSelected) {
      setActiveRect(rect);
    }
  }, [isSelected, rect, setActiveRect]);

  return (
    <Tab
      id={href}
      // TODO: Remove the use of `as`, which is deprecated. We can't pass a ref for an HTMLAnchorElement this way though...
      as={'a'}
      ref={ref}
      className={cn(TabStyles.TabLink, { [TabStyles.TabLink___isActive]: isSelected }, className)}
      {...props}
      render={<Link to={to} />}
    />
  );
}
```

**Comment by diegohaz** [comment]
Hi! If you pass `ref` to `render={<Link ref={ref} />}`, **you'll still get it merged**.

---

It's not automatically merged only if you use the function version of the `render` prop:

```jsx
render={(props) => <Link {...props} ref={mergeRefs(props.ref, ref)} />}
```

In this case, you can also use [Role](https://ariakit.org/components/role):

```jsx
<Tab
  render={(
    <Role.a
      ref={ref}
      render={(props) => <Link {...props} />}
    />
  )}
/>
```

This last method is still not documented, but that's the type-safe way to use the `render` *function prop* and still get props automatically merged.

**Reply by jasikpark** [reply]
Hmm what does `Role` get me? a typesafe way to ensure I'm accepting the right props?

**Reply by diegohaz** [reply, marked as answer]
Since you'd pass it directly to the `render` prop, it will automatically merge the props with the `Tab` component in a type-safe way:

```jsx
<Tab render={<Role.a ref={ref} />}>
```

That's equivalent to:

```jsx
<Tab render={<Link ref={ref} />}>
```

So you only need an intermediate `Role` if you're using the function version of `render`, which you are not (at least in your example):

```jsx
<Tab render={<Role.a ref={ref} render={(props) => <Link {...props} />} />}>
```

**Reply by jasikpark** [reply]
ok, thanks!
